docs: revise keyboard dongle guide (#3405)
zmkfirmware/zmk: docs/docs/hardware-integration/dongle.mdx

</ReadMore>

Depending on how the dongle is used, there are some additional [latency considerations](../features/split-keyboards.md#latency-considerations) to keep in mind.
The addition of the dongle adds an extra "hop" for the former central, increasing its latency to that of a peripheral. The other parts are unchanged latency-wise. There is also a commonly occurring case where the peripherals benefit.
Assuming the dongle is connected to USB and the former central would have been connected via bluetooth to the host if the dongle wasn't present:

- The former central will have its latency increased by about 1ms from the extra USB "hop"
- The other parts will have their average latency _decreased_ by 6.5ms from the replacement of a BLE "hop" with a USB "hop".

As a result, for this common use case the average latency of the keyboard decreases.

## Adding a Dongle

We'll add a new part with no keys as the central part, then switch the role of existing keyboard parts to peripheral.

This guide assumes that you are adding a dongle to an existing keyboard definition, with the keyboard boards and/or shields defined in the main ZMK repository or external modules.

> The dongle shield is defined in your personal ZMK config, no change is needed to the existing keyboard boards and/or shields.

Reviewer: This needs some sort of introductory phrase to make it read more natural. Maybe a reorganization like this:

> This guide describes a dongle setup where we add the dongle as a new part with no keys with the central role, then switch the role of existing keyboard parts to peripheral.
> It assumes that you are adding a dongle to an existing keyboard definition, with the keyboard boards and/or shields defined in the main ZMK repository or external modules.
> The dongle shield is defined in your personal ZMK config, no change is needed to the existing keyboard boards and/or shields.

If you are creating a new keyboard, the dongle shield can be defined together with other parts of the keyboard.

> Follow the [new keyboard shield guide](../hardware-integration/new-shield.mdx) and repeat the steps for "right" for each peripheral part of the keyboard.

Reviewer: The newly added "split keyboards with more than two parts" expanders cover the repeating remark, so we could probably omit that.

To add a dongle, you will create a simplified form of a [new shield](new-shield.mdx).
The approach described below assumes the dongle will not have any keys assigned to itself, so it will not work for that scenario.
If you want to understand the details of how it all works better, please read through the [new shield guide](new-shield.mdx).
This guide assumes that you are adding a dongle to an existing keyboard definition, with the keyboard boards and/or shields defined in the main ZMK repository or external modules.
The dongle shield is defined in your personal ZMK config, no change is needed to the existing keyboard boards and/or shields.

As there are a very large number of possible devices that could be used as a dongle, you will be defining your dongle as a personal shield intended for your exclusive use.
If you are creating a new keyboard, the dongle shield can be defined together with other parts of the keyboard.
Follow the [new keyboard shield guide](../hardware-integration/new-shield.mdx) and repeat the steps for "right" for each peripheral part of the keyboard.

> Prior to adding a dongle to your keyboard, please test its functionality without a dongle. The below guide will assume that your keyboard is named `my_keyboard`, replace accordingly.

Reviewer: Why remove the `my_keyboard` remark?

Author: Instead of:

- ... assume that your keyboard is named `my_keyboard`, replace accordingly
- Create a subdirectory called `my_keyboard`

Now it's worded:

- create a subdirectory with your keyboard's name
- for example `my_keyboard`

So I feel this is not needed anymore.

Reviewer: I see it now, mentioned under the Dongle Folder section. On second thought, maybe it'd be better to explicitly introduce this at the beginning? It bothers me that we aren't properly distinguishing keyboard name from shield/part names (not new with this PR), which also came up in the other comments I left. Something like:

> This guide describes creating a shield for a keyboard, named `my_keyboard` in below examples. The corresponding shield name is the same, unless the keyboard is a split keyboard where it will have two "part" shields named `my_keyboard_left` and `my_keyboard_right`.

Author: Added a new sentence about naming under the Dongle Folder section and removed the one under Kconfig.shield.

I skipped the part about the name of other parts, IMHO it sounds a bit too wordy and I had to read it twice to parse the meaning. Hopefully the current wording in the latest commit is clear enough.

After reading <https://docs.zephyrproject.org/latest/hardware/porting/shields.html#shield-variants> I do agree the current shield name terminology usage is ambiguous, but that would be addressed separately.

> Next, create a folder with the shield name of your keyboard under `zmk-config/boards/shields`, for example `zmk-config/boards/shields/my_keyboard`. This folder will contain the devicetree overlay and Kconfig files for your dongle.

Reviewer: ```suggestion
Next, create a folder with the name of your keyboard under `zmk-config/boards/shields`, for example `zmk-config/boards/shields/my_keyboard`. This folder will contain the devicetree overlay and Kconfig files for your dongle.
```

#### Kconfig.shield

Make a file called `Kconfig.shield`, if one does not exist already. Add the following lines to it, replacing `SHIELD_MY_KEYBOARD_DONGLE` and `my_keyboard_dongle` according to your keyboard:

```kconfig title="zmk-config/boards/shields/my_keyboard/Kconfig.shield"
# No whitespace after the comma or in the keyboard name!
config SHIELD_MY_KEYBOARD_DONGLE
    def_bool $(shields_list_contains,my_keyboard_dongle)
```

> To use the same keymap file, the dongle shield name should match other parts of the keyboard, e.g. `my_keyboard_dongle` if other parts are named `my_keyboard` (for unibody) or `my_keyboard_left` (for split).

Reviewer: This actually goes by the containing folder's name, which is the "keyboard name." The names of the parent folders of the main shield/board definition for the keyboard and the dongle's should match

Author: Looking at <https://github.com/zmkfirmware/zmk/blob/main/app/boards/post_boards_shields.cmake>, I don't think the folder name is used anywhere. On line 84, `${SHIELD_DIR}` is part of the `keymap_prefix` `foreach` but I searched both ZMK and Zephyr and I can't find where it's defined.

Reviewer: The shield_dir_name extracted in line 16, 40 etc. is the folder name of the shield, and I think the `SHIELD_DIR_${s}` it is extracted from is defined in https://github.com/zmkfirmware/zephyr/blob/v4.1.0%2Bzmk-fixes/cmake/modules/shields.cmake#L70 as the folder path containing the shield.overlay (for a shield example, also applies to boards). shield_dir_name in turn gets saved in shield_candidate_names which gets used for .keymap lookup etc.

};
```

#### Matrix Transform

Next, find all matrix transforms for your keyboard and copy them to the dongle shield. Look for a devicetree node with `compatible = "zmk,matrix-transform";` in the devicetree files for your keyboard.

<ReadMore>

>   [Devicetree Overview](../development/devicetree.md) explains the devicetree

Reviewer: ```suggestion
<ReadMore>
  See [Devicetree Overview](../development/devicetree.md) for details on the devicetree format.
```
(Dumb GH broke multiline suggestions, sorry)

>   format.

Reviewer: ```suggestion
```

};
```

If the `kscan` property is set on the physical layout node, remove it so the dongle uses the mock kscan instead.

```diff
 / {
     physical_layout0: physical_layout_0 {
         compatible = "zmk,physical-layout";
-        kscan = &kscan0;
         display-name = "Default Layout";
         transform = <&default_transform>;
         keys = <...>; // Long list of key positions, optional
     };
};
```

```diff
 &physical_layout0 {
-    kscan = &kscan0;
     transform = <&default_transform>;
 };
```

</TabItem>
<TabItem value="base">
If your keyboard has a physical layout defined in the same file as the matrix transform and chosen node, copy the physical layout node into your dongle shield's overlay.

> Silimar to the previous case with a dedicated physical layout file, make sure the matrix transform is assigned to it, and select it in the `chosen` node just like the other parts of the keyboard.

Reviewer: ```suggestion
Similar to the previous case with a dedicated physical layout file, make sure the matrix transform is assigned to it, and select it in the `chosen` node just like the other parts of the keyboard.
```

        compatible = "zmk,physical-layout";
        display-name = "Default Layout";
        transform = <&default_transform>;
        keys = <...>; // Long list of key positions, optional
    };
};
```

If the `kscan` property is set on the physical layout node, remove it so the dongle uses the mock kscan instead.

```diff
 / {
     physical_layout0: physical_layout_0 {
         compatible = "zmk,physical-layout";
-        kscan = &kscan0;
         display-name = "Default Layout";
         transform = <&default_transform>;
         keys = <...>; // Long list of key positions, optional
     };

> };

Reviewer: ```suggestion
 };
```

```

You can then flash the firmware to your device as detailed in our [user setup](../user-setup.mdx#install-the-firmware) and [split keyboard](../features/split-keyboards.md#building-and-flashing-firmware) pages.

:::warning
Before flashing your new firmware, you need to flash `settings_reset` [firmware](../troubleshooting/connection-issues.mdx#building-a-reset-firmware) on all devices to ensure they can pair to each other.
Before flashing your new firmware, you need to flash [`settings_reset` firmware](../troubleshooting/connection-issues.mdx#building-a-reset-firmware) on all devices to clear any previous bonding information.
Flashing standard ZMK firmware on a device will not clear any previous settings.
:::

To use your dongled keyboard with [ZMK Studio](../features/studio.md), apply the instructions for [building with Studio](../features/studio.md#building) to the dongle.

If you ever want to "undongle" your keyboard, simply remove these CMake arguments and flash the resulting firmware (after a `settings_reset`).

[^1]: If you have a custom dongle that uses an onboard MCU, then you will need to take a slightly different approach that isn't currently documented.

> [^2]: If you are building locally, you can append these flags to the end of the [build command](../development/local-toolchain/build-flash.mdx#cmake-arguments).

Reviewer: This might bo good to keep somewhere.

> If you ever want to "undongle" your keyboard, simply remove these CMake arguments, do the settings reset procedure, and flash the newly built firmware to your devices.

Reviewer: ```suggestion
If you ever want to "undongle" your keyboard, simply remove these CMake arguments, perform the settings reset procedure, and flash the newly built firmware to your devices.
```
